pandas/core/sparse/dtype.py

```diff
@@ -0,0 +1,186 @@
+        # __eq__, so we explicitly do it here.
+        return super(SparseDtype, self).__hash__()
+
+    def __eq__(self, other):
+        # We have to override __eq__ to handle NA values in _metadata.
+        # The base class does simple == checks, which fail for NA.
+        if isinstance(other, compat.string_types):
+            try:
+                other = self.construct_from_string(other)
+            except TypeError:
+                return False
+
+        if isinstance(other, type(self)):
+            subtype = self.subtype == other.subtype
+            if self._is_na_fill_value:
+                # this case is complicated by two things:
+                # SparseDtype(float, float(nan)) == SparseDtype(float, np.nan)
+                # SparseDtype(float, np.nan)     != SparseDtype(float, pd.NaT)
+                # i.e. we want to treat any floating-point NaN as equal, but
+                # not a floating-point NaN and a datetime NaT.
+                fill_value = (
+                    other._is_na_fill_value and
+                    isinstance(self.fill_value, type(other.fill_value)) or
+                    isinstance(other.fill_value, type(self.fill_value))
+                )
+            else:
+                fill_value = self.fill_value == other.fill_value
+
+            return subtype and fill_value
+        return False
+
+    @property
+    def fill_value(self):
+        """
+        The fill value of the array.
+
+        Converting the SparseArray to a dense ndarray will fill the
+        array with this value.
+
+        .. warning::
+
+           It's possible to end up with a SparseArray that has ``fill_value``
+           values in ``sp_values``. This can occur, for example, when setting
+           ``SparseArray.fill_value`` directly.
+        """
+        return self._fill_value
+
+    @property
+    def _is_na_fill_value(self):
+        from pandas.core.dtypes.missing import isna
+        return isna(self.fill_value)
+
+    @property
+    def _is_numeric(self):
+        from pandas.core.dtypes.common import is_object_dtype
+        return not is_object_dtype(self.subtype)
+
+    @property
+    def _is_boolean(self):
+        from pandas.core.dtypes.common import is_bool_dtype
+        return is_bool_dtype(self.subtype)
+
+    @property
+    def kind(self):
+        return self.subtype.kind
+
+    @property
+    def type(self):
+        return self.subtype.type
+
+    @property
+    def subtype(self):
+        return self._dtype
+
+    @property
+    def name(self):
+        return 'Sparse[{}]'.format(self.subtype.name)
+
+    def __repr__(self):
+        return 'Sparse[{},{}]'.format(self.subtype.name, self.fill_value)
+
+    @classmethod
+    def construct_array_type(cls):
+        from .array import SparseArray
+        return SparseArray
+
+    @classmethod
+    def construct_from_string(cls, string):
+        msg = "Could not construct SparseDtype from '{}'".format(string)
```

Comment:
I think it might be useful here to give a custom error message that passing a fill value is not supported.

Reply:
Hmm, what do you mean error message? I think we want to handle `'Spares[int, 0]'`, which I think means we have to treat the following the same

```
In [2]: pd.SparseDtype.construct_from_string("Sparse[int, 0]")
Out[2]: Sparse[int64, 0]

In [3]: pd.SparseDtype.construct_from_string("Sparse[int, 1]")
Out[3]: Sparse[int64, 0]
```

I will document that the `fill_value` is ignored though.

Reply:
To clarify, I made my comment in the light of that currently parsing a string repr with a fill_value errors. But if you actually want to support that, great, then a better error message is not needed :)

Reply:
Alrighty, resolving this then :)